firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,185 @@
+import android.util.SparseIntArray;
+import androidx.core.app.FrameMetricsAggregator;
+import androidx.fragment.app.Fragment;
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Optional;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ *
+ * <p>IMPORTANT: each recorder holds a reference to an Activity, so it is very important to
+ * dereference each recorder at or before its Activity's onDestroy. Similar for Fragments.
+ *
+ * <p>Relationship of Fragment recording to Activity recording is like a stopwatch. The stopwatch
+ * itself is for the Activity, Fragments traces are laps in the stopwatch. Stopwatch can only ever
+ * be for the Activity by-definition because "frames" refer to a frame for a window, and
+ * FrameMetrics can only come from an Activity's Window. Fragment traces are laps in the stopwatch,
+ * because the frame metrics data is still for the Activity window, fragment traces are just
+ * intervals in the Activity frames recording that has the name "fragment X" attached to them.
+ */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+
+  private final Activity activity;
+  private final FrameMetricsAggregator frameMetricsAggregator;
+  private final Map<Fragment, PerfFrameMetrics> fragmentSnapshotMap;
+
+  private boolean isRecording = false;
+
+  /**
+   * Creates a recorder for a specific activity.
+   *
+   * @param activity the activity that the recorder is collecting data from.
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new HashMap<>());
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(
+      Activity activity,
+      FrameMetricsAggregator frameMetricsAggregator,
+      Map<Fragment, PerfFrameMetrics> subTraceMap) {
+    this.activity = activity;
+    this.frameMetricsAggregator = frameMetricsAggregator;
+    this.fragmentSnapshotMap = subTraceMap;
+  }
+
+  /** Starts recording FrameMetrics for the activity window. */
+  public void start() {
+    if (isRecording) {
+      logger.debug(
+          "FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    frameMetricsAggregator.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window.
+   *
+   * @return FrameMetrics accumulated during the current recording.
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.debug("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    if (!fragmentSnapshotMap.isEmpty()) {
```

Comment:
Do we need to clear the `fragmentSnapshotMap`? If not, when we start another activity and want to call `startFragment(Fragment fragment)`, the map could be non-empty. Therefore, we will not be able to start certain fragment.

Reply:
Good point! Added `fragmentSnapshotMap.clear()`.

Reply:
Can you add a test that would trigger this behavior so we don't accidentally reintroduce the bug in the future?

Reply:
Done!